File: bridge/src/discovery.js
// discovery.js — mDNS advertisement so the board finds the Mac (SPEC §4.6).
//
// Advertises `_ytmboard._tcp` on the board-facing WS port. The board browses for
// this service and connects to the advertised host:port — no hard-coded IP.
import { Bonjour } from "bonjour-service";
import { BOARD_PORT } from "./config.js";

const SERVICE_TYPE = "ytmboard"; // -> _ytmboard._tcp
const SERVICE_NAME = "YT Music board bridge";

export class Discovery {
  constructor() {
    this.bonjour = null;
    this.service = null;
  }

  start() {
    this.bonjour = new Bonjour();
    this.service = this.bonjour.publish({
      name: SERVICE_NAME,
      type: SERVICE_TYPE,
      protocol: "tcp",
      port: BOARD_PORT,
      // Board reads these to know how to talk to us (plain WS, protocol v1).
      txt: { proto: "ws", path: "/", v: "1" },
    });
    console.log(`[mdns] advertising _${SERVICE_TYPE}._tcp on :${BOARD_PORT}`);
    return this.service;
  }

  stop() {
    // Send goodbye packets so boards drop us promptly, then tear down.
    try {
      this.bonjour?.unpublishAll(() => this.bonjour?.destroy());
    } catch {
      this.bonjour?.destroy?.();
    }
  }
}

File: app/bridge-core/src/discovery.rs
//! mDNS advertisement so the board finds the host (ports discovery.js).
//! Advertises `_ytmboard._tcp` on the board port; the board browses for this
//! service and connects to the advertised host:port — no hard-coded IP.
use crate::config::board_port;
use anyhow::{Context, Result};
use mdns_sd::{ServiceDaemon, ServiceInfo};

const SERVICE_TYPE: &str = "_ytmboard._tcp.local.";
const SERVICE_NAME: &str = "YT Music board bridge";

pub struct Discovery {
    daemon: ServiceDaemon,
}

pub fn start() -> Result<Discovery> {
    let daemon = ServiceDaemon::new().context("mdns ServiceDaemon::new")?;
    let service = ServiceInfo::new(
        SERVICE_TYPE,
        SERVICE_NAME,
        "ytmboard.local.",
        "",
        board_port(),
        &[("proto", "ws"), ("path", "/"), ("v", "1")][..],
    )
    .context("mdns ServiceInfo::new")?
    .enable_addr_auto();
    daemon.register(service).context("mdns register")?;
    tracing::info!("[mdns] advertising {SERVICE_TYPE} on :{}", board_port());
    Ok(Discovery { daemon })
}

impl Discovery {
    pub fn stop(self) {
        // Send goodbye packets so boards drop us promptly, then shut down.
        let _ = self.daemon.unregister(&format!("{SERVICE_NAME}.{SERVICE_TYPE}"));
        let _ = self.daemon.shutdown();
    }
}
